File: TE_1 SEM/AI PRACTICAL/A_2.cpp
#include <bits/stdc++.h>
using namespace std;

// Define Pair as std::pair<int, int> for clarity
typedef pair<int, int> Pair;

const int ROW = 5;
const int COL = 5;

struct Cell {
    int parentX, parentY;
    double f, g, h;
};

// A utility function to check if a given cell (row, col) is a valid cell
bool isValid(int row, int col) {
    return (row >= 0) && (row < ROW) && (col >= 0) && (col < COL);
}

// Check if the cell is an unblocked cell
bool isUnblocked(int grid[][COL], int row, int col) {
    return (grid[row][col] == 1);
}

// Check if cell (x, y) is the destination cell
bool isDestination(int x, int y, Pair dest) {
    return (x == dest.first && y == dest.second);
}

// Calculate the 'h' heuristic, which is the estimated cost to reach destination
double calculateHValue(int x, int y, Pair dest) {
    return abs(x - dest.first) + abs(y - dest.second);
}

// A* Search Algorithm
void aStarSearch(int grid[][COL], Pair start, Pair dest) {
    if (!isValid(start.first, start.second) || !isValid(dest.first, dest.second) ||
        !isUnblocked(grid, start.first, start.second) || !isUnblocked(grid, dest.first, dest.second)) {
        cout << "Invalid start or destination" << endl;
        return;
    }

    if (isDestination(start.first, start.second, dest)) {
        cout << "Already at the destination" << endl;
        return;
    }

    bool closedList[ROW][COL];
    memset(closedList, false, sizeof(closedList));

    Cell cellDetails[ROW][COL];

    for (int i = 0; i < ROW; i++) {
        for (int j = 0; j < COL; j++) {
            cellDetails[i][j].f = FLT_MAX;
            cellDetails[i][j].g = FLT_MAX;
            cellDetails[i][j].h = FLT_MAX;
            cellDetails[i][j].parentX = -1;
            cellDetails[i][j].parentY = -1;
        }
    }

    int x = start.first, y = start.second;
    cellDetails[x][y] = { x, y, 0.0, 0.0, 0.0 };

    set<pair<double, Pair>> openList;
    openList.insert(make_pair(0.0, make_pair(x, y)));
    bool foundDest = false;

    while (!openList.empty()) {
        pair<double, Pair> current = *openList.begin();
        openList.erase(openList.begin());

        x = current.second.first;
        y = current.second.second;
        closedList[x][y] = true;

        // Generate the 4 possible moves (up, down, left, right)
        int dx[] = { -1, 1, 0, 0 };
        int dy[] = { 0, 0, -1, 1 };

        for (int i = 0; i < 4; i++) {
            int newX = x + dx[i], newY = y + dy[i];

            if (isValid(newX, newY)) {
                if (isDestination(newX, newY, dest)) {
                    cellDetails[newX][newY].parentX = x;
                    cellDetails[newX][newY].parentY = y;
                    cout << "Destination found" << endl;
                    foundDest = true;

                    stack<Pair> path;
                    Pair cell = make_pair(newX, newY);
                    while (!(cell.first == start.first && cell.second == start.second)) {
                        path.push(cell);
                        int px = cellDetails[cell.first][cell.second].parentX;
                        int py = cellDetails[cell.first][cell.second].parentY;
                        cell = make_pair(px, py);
                    }
                    path.push(start);

                    while (!path.empty()) {
                        Pair p = path.top();
                        path.pop();
                        cout << "(" << p.first << ", " << p.second << ") ";
                    }
                    cout << endl;
                    return;
                } else if (!closedList[newX][newY] && isUnblocked(grid, newX, newY)) {
                    double gNew = cellDetails[x][y].g + 1.0;
                    double hNew = calculateHValue(newX, newY, dest);
                    double fNew = gNew + hNew;

                    if (cellDetails[newX][newY].f == FLT_MAX || cellDetails[newX][newY].f > fNew) {
                        openList.insert(make_pair(fNew, make_pair(newX, newY)));
                        cellDetails[newX][newY] = { x, y, fNew, gNew, hNew };
                    }
                }
            }
        }
    }

    if (!foundDest) {
        cout << "Failed to find the Destination" << endl;
    }
}

// Driver code
int main() {
    int grid[ROW][COL] = {
        { 1, 1, 1, 1, 1 },
        { 1, 0, 1, 0, 1 },
        { 1, 1, 1, 1, 1 },
        { 0, 0, 1, 0, 1 },
        { 1, 1, 1, 1, 1 }
    };

    Pair start = make_pair(0, 0);  // Starting point
    Pair dest = make_pair(4, 4);   // Destination

    aStarSearch(grid, start, dest);

    return 0;
}

File: TE_1 SEM/AI PRACTICAL/A_2.py
# Define constants
ROW = 5
COL = 5

# Define a cell with parent coordinates and cost values
class Cell:
    def __init__(self, parentX, parentY, f=float('inf'), g=float('inf'), h=float('inf')):
        self.parentX = parentX
        self.parentY = parentY
        self.f = f
        self.g = g
        self.h = h

# Check if a cell is valid (within grid bounds)
def is_valid(row, col):
    return 0 <= row < ROW and 0 <= col < COL

# Check if a cell is unblocked (value 1 means unblocked, 0 means blocked)
def is_unblocked(grid, row, col):
    return grid[row][col] == 1

# Check if the cell is the destination
def is_destination(x, y, dest):
    return x == dest[0] and y == dest[1]

# Calculate the Manhattan Distance heuristic
def calculate_h_value(x, y, dest):
    return abs(x - dest[0]) + abs(y - dest[1])

# Find the smallest f-value in open_list
def find_min_f(open_list):
    min_f = float('inf')
    min_node = None
    for cell in open_list:
        if cell[0] < min_f:
            min_f = cell[0]
            min_node = cell
    return min_node

# A* Search Algorithm
def a_star_search(grid, start, dest):
    if not is_valid(start[0], start[1]) or not is_valid(dest[0], dest[1]) or \
       not is_unblocked(grid, start[0], start[1]) or not is_unblocked(grid, dest[0], dest[1]):
        print("Invalid start or destination")
        return
    
    if is_destination(start[0], start[1], dest):
        print("Already at the destination")
        return
    
    closed_list = [[False] * COL for _ in range(ROW)]
    cell_details = [[Cell(-1, -1) for _ in range(COL)] for _ in range(ROW)]

    # Initialize the start cell
    x, y = start
    cell_details[x][y] = Cell(x, y, f=0.0, g=0.0, h=0.0)

    open_list = [(0.0, (x, y))]
    found_dest = False

    # Directions for moving up, down, left, and right
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

    while open_list:
        # Find the cell with the lowest f-value
        current = find_min_f(open_list)
        open_list.remove(current)
        x, y = current[1]
        closed_list[x][y] = True

        # Check each possible move (up, down, left, right)
        for dx, dy in directions:
            new_x, new_y = x + dx, y + dy

            if is_valid(new_x, new_y):
                if is_destination(new_x, new_y, dest):
                    cell_details[new_x][new_y].parentX = x
                    cell_details[new_x][new_y].parentY = y
                    print("Destination found")
                    found_dest = True

                    # Trace the path
                    path = []
                    cell = (new_x, new_y)
                    while cell != start:
                        path.append(cell)
                        px, py = cell_details[cell[0]][cell[1]].parentX, cell_details[cell[0]][cell[1]].parentY
                        cell = (px, py)
                    path.append(start)
                    path.reverse()
                    print("Path:", path)
                    return
                
                elif not closed_list[new_x][new_y] and is_unblocked(grid, new_x, new_y):
                    g_new = cell_details[x][y].g + 1.0
                    h_new = calculate_h_value(new_x, new_y, dest)
                    f_new = g_new + h_new

                    if cell_details[new_x][new_y].f == float('inf') or cell_details[new_x][new_y].f > f_new:
                        open_list.append((f_new, (new_x, new_y)))
                        cell_details[new_x][new_y] = Cell(x, y, f_new, g_new, h_new)

    if not found_dest:
        print("Failed to find the destination")

# Driver code
if __name__ == "__main__":
    grid = [
        [1, 1, 1, 1, 1],
        [1, 0, 1, 0, 1],
        [1, 1, 1, 1, 1],
        [0, 0, 1, 0, 1],
        [1, 1, 1, 1, 1]
    ]

    start = (0, 0)  # Starting point
    dest = (4, 4)   # Destination

    a_star_search(grid, start, dest)
